specs/latest/1.0/index.html

```diff
@@ -730,6 +730,12 @@ <h3>Types</h3>
 // The power preference settings are documented in the WebGLContextAttributes
 // section of the specification.
 enum WebGLPowerPreference { "default", "low-power", "high-performance" };
+
+enum WebGLPixelFormat {
+    "RGBA8",
+    "SRGB8_ALPHA8",
```

Comment:
I think there is value in including RGB10_A2 as an option here. Few ideas of where this will become useful
- this lets us do full rec2020 with 32-bits-per-pixel (not blowing up to 64-bits-per-pixel)
- medical imaging applications which demand high resolution of greyscale (IEEE 754 has enough, but maybe they would be particular about fixed-point?)
- if we end up adding rec2020-pq as a color space at some point in the future, this would be the ideal format for it

That said, this is a perfectly good starting set to have.

Reply:
Would Rec. 2020 with `RGB10_A2` imply applying non-linear ITU transfer using GLSL when writing to `gl_FragColor` (like it happens with `RGBA8` and sRGB today)?

Reply:
Long answer to avoid confusing the issue.

The value written to the back buffer is in the canvas' color space, whatever that space is.

The value assigned to gl_FragColor may not be the same as the value that is written to the backbuffer. 
- In particular, if the backbuffer has sRGB framebuffer encoding enabled (which is implicitly true if the format specified here is SRGB8_ALPHA -- on desktop platforms it's true when FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING is SRGB and FRAMEBUFFER_SRGB is enabled), then the linear-to-sRGB-encoded-transform function is applied before writing the value.
- In that case, gl_FragColor is in a "linearized" version of the canvas' color space.
  - This is true for "srgb" and "display-p3" color spaces, because their transfer function 
  - For something like "rec2020" or "rec2020-pq", this ends up being a well-defined, but not very useful space (not linear, not standard, not ... anything, just the result of a bunch of formulas that aren't usually put together)

So, regarding ``RGBA10_A2``
- no hardware supports any framebuffer color encoding functions for 10-bit formats
  - not sRGB, which we do have for 8-bit formats
  - and not PQ, which doesn't exist in any hardware (but hey, maybe it would be a nice feature request!)
- so the values written would be "values in [0, 1] in the rec2020 color space"

Does that answer the question?

BTW, the terminology in KHR_gl_colorspace is really bad and confusing (they use the word "color space" for things that aren't color spaces, and then when they do talk about real color spaces, they invert the meanings of terms). So beware of that.

Reply:
> Does that answer the question?

BT.2020 defines its own legacy OETF for 10- and 12-bit quantization. Additionally, BT.2100 provides HLG and PQ transfers, that are likely better suited for HDR content.

Writing real linear values `[0.0 .. 1.0]` to `RGBA10_A2` would likely lead to losing some perceptual precision compared to pre-applying some known to compositor non-linear transfer in the fragment shader.

Reply:
Yeah, srgb-encoding is primarily to address the banding you'd get if you tried to store linear values directly into unorm8. It solves this by applying the TF, such that any unorm8 quantization banding is now equally spread throughout the more-perceptually-flat E side of the OETF. With 10bpc, you don't have to worry about banding as much, I suppose, and you could just tag the canvas as e.g. rec2020-linear. (Or srgb-linear! Or whatever!) It's on the app to tag correctly based on the output of their rendering, but I think with this, the major pieces of the puzzle are ready.

Would prefer to handle this in a follow-up, but it's a good idea.

Reply:
@lexaknyazev Yes, but c'est la vie! That's on the app to do, since we can (in particular) support linearization for blending there.

Reply:
Ah, I completely forgot about blending-in-backbuffer scenario.

> With 10bpc, you don't have to worry about banding as much

For sRGB or P3 gamut, maybe. For a much wider BT.2020, there is a strong suggestion in BT.2246 to use at least _10 bits with non-linear transfer_.

Reply:
> no hardware supports any framebuffer color encoding functions for 10-bit formats

FWIW, modern Apple GPUs have this: https://developer.apple.com/documentation/metal/mtlpixelformat/mtlpixelformatbgr10_xr_srgb

Reply:
Wow, this is the first I'd ever seen such a thing! Maybe it will become a thing! I don't know of any standard BT2020-primaries + sRGB-transfer-function color space that is common (which is where this would shine).

Reply:
>>With 10bpc, you don't have to worry about banding as much

>For sRGB or P3 gamut, maybe.

Actually, it may be not enough sometimes even for standard range content: the linear precision of 8-bit sRGB roughly varies from 12 to 7 bits. Also applications would have to be aware of blending with only 2 bits of alpha.

So my point here is that using `RGBA10_A2` as a backbuffer storage is a very niche option with non-obvious tradeoffs.

Reply:
> if we end up adding rec2020-pq as a color space at some point in the future

How would blending work there without platform-provided linearization? 10-bit PQ transfer is even more aggressive than typical sRGB-style curves as PQ's linear precision may achieve 28 bits for values near zero.

![image](https://user-images.githubusercontent.com/3479527/109382865-6bc38700-78fc-11eb-9fb0-7c72aff28369.png)

Reply:
Looking forward I think most of fancier spaces/tfs will just want rgba16f, but we'll keep an eye out for other options.